### PR TITLE
feat(persons-on-events): async migration for persons-on-events

### PR DESCRIPTION
**Pull request by macobo**
## Problem

Ticket: https://github.com/PostHog/posthog/issues/10561

Follow-up to: https://github.com/PostHog/posthog/pull/10699

## Changes

This PR:
- Adds the new async migration
- Updates some existing testing-related or async migration utils
- Compresses person and group property columns with ZSTD(3) by default

Not in scope:
- Parameterizing in UI
- Making this runnable on self-hosted

## How did you test this code?

See tests.

### CODE REVIEW

File: posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+            "group0_properties",
+            "group1_properties",
+            "group2_properties",
+            "group3_properties",
+            "group4_properties",
+        ]
+        for column in columns:
+            execute_op_clickhouse(
+                query_id=query_id,
+                sql=f"ALTER TABLE {EVENTS_DATA_TABLE()} MODIFY COLUMN {column} VARCHAR Codec({codec})",
```
Comment (macobo) [inline posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py:307]: ON CLUSTER!

File: posthog/async_migrations/disk_util.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from posthog.client import sync_execute
+from posthog.settings import CLICKHOUSE_DATABASE
+
+
+def analyze_enough_disk_space_free_for_table(table_name: str, required_ratio: float):
```
Comment (yakkomajuri) [inline posthog/async_migrations/disk_util.py:5]: nice util

File: posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+                    CREATE DICTIONARY IF NOT EXISTS person_distinct_id2_dict {{on_cluster_clause}}
+                    (
+                        team_id Int64,
+                        distinct_id String,
+                        person_id UUID
+                    )
+                    PRIMARY KEY team_id, distinct_id
+                    SOURCE(CLICKHOUSE(TABLE {TEMPORARY_PDI2_TABLE_NAME} {self._dictionary_connection_string()}))
+                    LAYOUT(complex_key_cache(size_in_cells 50000000 max_threads_for_updates 6 allow_read_expired_keys 1))
+                    Lifetime(60000)
```
Comment (yakkomajuri) [inline posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py:222]: q: just curious, did you ever experiment with different lifetimes?

**Review by yakkomajuri — COMMENTED** [review]
sending the review out before looking over the tests to maybe speed you up a bit

overall looks good, a few comments and questions

File: posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py (hunk 3 above)
Comment (macobo) [reply posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py:222]: Nope. It didn't matter for the tests I ran w/ 200M events. What are your thoughts on this, how could we make use of this?

File: posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        sleep_until_finished("events table backill", lambda: self._count_running_mutations() > 0)
+
+    def _count_running_mutations(self):
+        return sync_execute(
+            """
+            SELECT count()
+            FROM clusterAllReplicas(%(cluster)s, system, 'mutations')
+            WHERE not is_done AND command LIKE %(pattern)s
+            """,
+            {"cluster": settings.CLICKHOUSE_CLUSTER, "pattern": "%person_properties = dictGetString%",},
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py:321]: It reformats for the mutations table but let's align this with the code. Good note

File: posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+            f"DROP DICTIONARY IF EXISTS person_dict {{on_cluster_clause}}",
+            f"DROP DICTIONARY IF EXISTS person_distinct_id2_dict {{on_cluster_clause}}",
+            f"DROP DICTIONARY IF EXISTS groups_dict {{on_cluster_clause}}",
+            f"DROP TABLE IF EXISTS {TEMPORARY_PERSONS_TABLE_NAME} {{on_cluster_clause}}",
+            f"DROP TABLE IF EXISTS {TEMPORARY_PDI2_TABLE_NAME} {{on_cluster_clause}}",
+            f"DROP TABLE IF EXISTS {TEMPORARY_GROUPS_TABLE_NAME} {{on_cluster_clause}}",
+        ]
+        for query in queries:
+            execute_op_clickhouse(query_id=query_id, sql=query, per_shard=True)
+
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py:335]: Basically yeah, decided that estimating progress here is annoying. Parts to do might be useful if we decide to override this, but getting the total number of parts is slightly annoying in a live env

File: posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    depends_on = "0005_person_replacing_by_version"
+
+    def precheck(self):
+        if not settings.MULTI_TENANCY:
+            return False, "This async migration is not yet ready for self-hosted users"
+
+        return analyze_enough_disk_space_free_for_table(EVENTS_DATA_TABLE(), required_ratio=2.0)
+
+    def is_required(self) -> bool:
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py:70]: 100% 

Note I have considered switching this to a count blank person_id columns check, but let's iterate in a follow-up

File: posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+                    ORDER BY (team_id, group_type_index, group_key)
+                    SETTINGS index_granularity = 128
+                """,
+                rollback=f"DROP TABLE IF EXISTS {TEMPORARY_GROUPS_TABLE_NAME} {{on_cluster_clause}}",
+                per_shard=True,
+            ),
+            AsyncMigrationOperationSQL(
+                sql=f"""
+                    ALTER TABLE {TEMPORARY_PERSONS_TABLE_NAME} {{on_cluster_clause}}
+                    REPLACE PARTITION tuple() FROM person
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py:125]: Only works since we don't partition this table :)

**Pull request merged**